cachi2/core/models/sbom.py

```diff
@@ -87,20 +86,624 @@ class Metadata(pydantic.BaseModel):
+            pRel = partial(SPDXRelation, spdxElementId=root_id, comment="", relationshipType=rtype)
+            for package in packages:
+                if package.SPDXID == "SPDXRef-DocumentRoot-File-":
+                    continue
+                relationships.append(pRel(relatedSpdxElement=package.SPDXID))
+            return relationships
+
+        def libs_to_packages(libraries: list[Component]) -> list[SPDXPackage]:
+            packages, annottr, now = [], "Tool: cachi2:jsonencoded", spdx_now()
+            args = dict(annotator=annottr, annotationDate=now, annotationType="OTHER")
+            pAnnotation = partial(SPDXPackageAnnotation, **args)
+
+            # noqa for trivial helpers.
+            mkcomm = lambda p: json.dumps(dict(name=f"{p.name}", value=f"{p.value}"))  # noqa: E731
+            hashdict = lambda c: dict(name=c.name, version=c.version, purl=c.purl)  # noqa: E731
+            erefbase = dict(referenceCategory="PACKAGE-MANAGER", referenceType="purl")
+            erefdict = lambda c: dict(referenceLocator=c.purl, **erefbase)  # noqa: E731
+
+            for component in libraries:
+                package_hash = SPDXPackage._calculate_package_hash_from_dict(hashdict(component))
+                packages.append(
+                    SPDXPackage(
+                        SPDXID=f"SPDXRef-Package-{component.name}-{component.version}-{package_hash}",
+                        name=component.name,
+                        versionInfo=component.version,
+                        externalRefs=[erefdict(component)],
+                        annotations=[pAnnotation(comment=mkcomm(p)) for p in component.properties],
+                    )
+                )
+            return packages
+
+        def files_to_packages(cydxfiles: list[Component]) -> list[SPDXFile]:
+            return [SPDXFile(SPDXID="SPDXRef-File-{f.name}", fileName=f.name) for f in cydxfiles]
+
+        # Main function body.
+        cydxfiles, libraries = partition_by(lambda c: c.type == "library", self.components)
+        # mypy is upset by patrition_by being broadly typed.
+        packages = [create_document_root()] + libs_to_packages(libraries)  # type: ignore
+        files = files_to_packages(cydxfiles)  # type: ignore
+        relationships = [create_root_relationship()] + link_to_root(packages) + link_to_root(files)
+        # noqa for a trivial helper.
+        creator = lambda tool: [f"Tool: {tool.name}", f"Organization: {tool.vendor}"]  # noqa: E731
+        return SPDXSbom(
+            packages=packages,
+            relationships=relationships,
+            files=files,
+            documentNamespace=doc_namespace,
+            creationInfo=SPDXCreationInfo(
+                creators=sum([creator(tool) for tool in self.metadata.tools], []),
+                created=spdx_now(),
+            ),
+        )
+
+
+class SPDXPackageExternalRefReferenceLocatorURI(pydantic.BaseModel):
```

Comment:
Nitpick: Maybe we could move these model classes to a separate `sbom-spdx.py` module, as this one here has already grown too large.

Reply:
It is doable, but will be a bit more complicated than that: both SBOM types need to be aware of each other, so yet another module would be needed to avoid imports from partially initialized modules.

Reply:
Seems like a major refactor would be needed, which I think would be benefitial, but out of scope for this PR.